### ISSUE TITLE
[MAINT] switch to a src layout to organize the tests

### ISSUE THREAD

**Issue by Remi-Gau** [nilearn/nilearn#3660]
**Edit**

This issue has been renamed as the topic of having __init__.py in the tests folder fits better within a more global scope of whether we would like to adopt a source layout.

---

Unless I am mistaken, is there a reason why the nilearn tests should be part of the package?

If not, I suspect that there is a bunch of __init__.py files that can be removed.

**Comment by bthirion** [comment]
I see that MNE has a bunch of __init__.py files in the test folders. Dunno why.

**Comment by ymzayek** [comment]
I guess it's because we follow this [test as part of application code](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#tests-as-part-of-application-code) setup. But there is a note below that section with this relevant part: 

> If you use one of the two recommended file system layouts above but leave away the __init__.py files from your directories, it should just work. From “inlined tests”, however, you will need to use absolute imports for getting at your application code.

So I guess they are not necessary if we're using absolute imports for our testing (IIUC).

Whether the tests should be part of the package is another question?

**Comment by jeromedockes** [comment]
as @ymzayek says the `__init__.py` files are there because the tests are part of the package, and get installed along with the rest of nilearn. the reason that is the case is that when scikit-learn and then nilearn (and MNE) were created, that was the normal way of laying out a python package and its tests. It is now common (and IMO desirable) to use the "src" [layout](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#tests-outside-application-code). we can discuss the pros and cons of making that change.

to summarize -- I don't think there is any reason to remove these `__init__.py` files

**Comment by jeromedockes** [comment]
with the old setuptools configuration (0.10.0 release), if the `__init__.py` files are missing the test packages do not get copied to the installation directory when installing nilearn, so for example `pytest --pyargs nilearn` does not run the tests
with the new configuration they do get installed, but as they are missing an `__init__.py` they are [implicit namespace packages](https://peps.python.org/pep-0420/), which is not necessarily a big problem but does not make much sense as we are not in the situation they are intended for:

> Namespace packages are a mechanism for splitting a single Python package across multiple directories on disk

you can check they are namespace packages by importing one of the packages missing the `__init__.py`: it has no `__file__`, its `__path__` is a `_NamespacePath`, its `__loader__` is a `_NamespaceLoader`

btw I noticed the installation of some of the test data got broken during the switch to hatchling: for example `nilearn/datasets/tests/data/craddock_2011_parcellations.tar.gz` does not get copied to the installation directory (I suspect that might be the case for some other .tar.gz files?)
if we don't decide to switch to an `src` layout (without installable tests), we should probably fix that

**Comment by jeromedockes** [comment]
as a result, I believe [the azure pipelines](https://github.com/nilearn/nilearn/blob/54f4ccda94c6b9c4e4389f0c48804738904f6218/azure-pipelines.yml#L42) tests only succeed by chance because (i) we run them from the root of the repository, which contains the `nilearn` source (a frequent source of problems...) _and_ (ii) we invoke pytest with `python -m pytest`, which adds the current directory to the path ([doc](https://docs.pytest.org/en/7.1.x/how-to/usage.html#calling-pytest-through-python-m-pytest)) . I haven't tried on azure but locally the tests pass with `python -m pytest --pyargs=nilearn` from the root of the repository, but fail due to the missing craddock file (I only ran the datasets tests) if invoking the pytest command directly or from another directory

**Comment by jeromedockes** [comment]
one (small) advantage of having the tests as part of the package we forgot to mention in the meeting is that users can run the tests after installing the package `pip install nilearn[test]` then `pytest --pyargs=nilearn`.

**Comment by Remi-Gau** [comment]
Renamed the issue.

Adding some links to explain what the src layout is:

- https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/
- https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#choosing-a-test-layout-import-rules
- https://www.youtube.com/watch?v=sW1qUZ_nSXk

**Comment by ymzayek** [comment]
> btw I noticed the installation of some of the test data got broken during the switch to hatchling: for example nilearn/datasets/tests/data/craddock_2011_parcellations.tar.gz does not get copied to the installation directory (I suspect that might be the case for some other .tar.gz files?)
if we don't decide to switch to an src layout (without installable tests), we should probably fix that

Thanks for catching that, let me check to see if I can fix it quickly. I think we should hold off on making a quick decision to switch to the new layout especially given that formatting and refactoring project has required moving things around so at least we can wait to finish that.

**Comment by Remi-Gau** [comment]
> I think we should hold off on making a quick decision to switch to the new layout

agreed I did set a low priority to this discussion: but at least we have a venue to talk about it

**Comment by jeromedockes** [comment]
after discussion with @ymzayek  in the drop-in hours we decided to remove .tar.gz from the .gitignore

**Comment by Remi-Gau** [comment]
May be good to move some tests out of the nilearn package: the last version (0.12.0) contains tests that compare the outcome of several plotting functions to some baseline figures. The baseline figures are included in the package which increased the size of the package by quite a bit.

![Image](https://github.com/user-attachments/assets/29f7cf49-18c1-433f-b86e-1ff9ab04571f)


![Image](https://github.com/user-attachments/assets/db5eb8de-58a8-45d1-b2a7-e4978108c7e9)


Would be worth to see if we can use extract some of the tests from the nilearn folder.

**Comment by Remi-Gau** [comment]
note that the change of layout would also simplify visualizing the 'real' architecture as pydeps includes all tests folder when plotting nilearn internal dependency grpah.

```
pydeps nilearn -v --noshow --only nilearn --max-module-depth 2 --exclude-exact nilearn.conftest nilearn.tests nilearn.input_data
```

now

<img width="1501" height="756" alt="Image" src="https://github.com/user-attachments/assets/668ab1c7-3068-4927-9787-3e91ed3f1d62" />

after removing all tests folder from nilearn/

<img width="1401" height="753" alt="Image" src="https://github.com/user-attachments/assets/f511edb8-3171-4587-9b76-6ad4cd1db98f" />
